### PR TITLE
feat: DeactivateDid transaction to mark DIDs as deactivated [DEV-1774]

### PR DESCRIPTION
**Pull request by khornonzon**

### CODE REVIEW

File: proto/cheqd/v1/tx.proto [hunk 1]
```diff
@@ -23,11 +24,38 @@ message MsgUpdateDid {
...
+
 message SignInfo {
   string verification_method_id = 1;
   string signature = 2;
 }
 
+// created DeactivateDid payload
+message MsgDeactivateDidPayload {
+  repeated string context = 1;
+  string id = 2;
```
Comment (Toktar) [inline proto/cheqd/v1/tx.proto:42]: Why this proto doesn't contain only `id` field?

File: proto/cheqd/v1/tx.proto [hunk 2]
```diff
@@ -23,11 +24,38 @@ message MsgUpdateDid {
...
+  repeated string capability_invocation = 7;
+  repeated string capability_delegation = 8;
+  repeated string key_agreement = 9;
+  repeated string also_known_as = 10;
+  repeated Service service = 11;
+}
+
+//created DeactivateDid response
+message MsgDeactivateDidResponse {
+  string id = 1;
```
Comment (Toktar) [inline proto/cheqd/v1/tx.proto:56]: Let's add a metadata and maybe a full DidDoc to a response

**Review by Toktar — CHANGES_REQUESTED** [review]
Could you please start with a Plan of Attack?

File: scripts/protocgen.sh [hunk 3]
```diff
@@ -16,17 +16,17 @@ proto_dirs=$(find . -path ./third_party -prune -o -name '*.proto' -print0 | xarg
 for dir in $proto_dirs; do
   # Generate protobuf bind
   # shellcheck disable=SC2046
-  buf protoc \
-  -I "proto" \
-  -I "$cosmos_sdk_dir/third_party/proto" \
-  -I "$cosmos_sdk_dir/proto" \
-  --gocosmos_out=plugins=interfacetype+grpc,\
-Mgoogle/protobuf/any.proto=github.com/cosmos/cosmos-sdk/codec/types:. \
+#    protoc \
```
Comment (Toktar) [inline scripts/protocgen.sh:19]: Could you please return these lines?

File: scripts/protocgen.sh [hunk 4]
```diff
@@ -16,17 +16,17 @@ proto_dirs=$(find . -path ./third_party -prune -o -name '*.proto' -print0 | xarg
...
+#   -I "proto" \
+#   -I "$cosmos_sdk_dir/third_party/proto" \
+#   -I "$cosmos_sdk_dir/proto" \
+#   --gocosmos_out=plugins=interfacetype+grpc,\
+# Mgoogle/protobuf/any.proto=github.com/cosmos/cosmos-sdk/codec/types:. \
   $(find "${dir}" -name '*.proto')
 
   # Generate grpc gateway
   # shellcheck disable=SC2046
-  buf protoc \
```
Comment (Toktar) [inline scripts/protocgen.sh:29]: Could you please return this?

File: x/cheqd/types/query.pb.gw.go [hunk 5]
```diff
@@ -181,7 +181,7 @@ func RegisterQueryHandlerClient(ctx context.Context, mux *runtime.ServeMux, clie
 }
 
 var (
-	pattern_Query_Did_0 = runtime.MustPattern(runtime.NewPattern(1, []int{2, 0, 2, 1, 2, 2, 1, 0, 4, 1, 5, 3}, []string{"cheqd", "v1", "did", "id"}, "", runtime.AssumeColonVerbOpt(false)))
+	pattern_Query_Did_0 = runtime.MustPattern(runtime.NewPattern(1, []int{2, 0, 2, 1, 2, 2, 1, 0, 4, 1, 5, 3}, []string{"cheqd", "v1", "did", "id"}, "", runtime.AssumeColonVerbOpt(true)))
```
Comment (Toktar) [inline x/cheqd/types/query.pb.gw.go:184]: Could you please return this?

File: x/cheqd/client/cli/tx_deactivate_did.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/cheqd/cheqd-node/x/cheqd/types"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/spf13/cobra"
+)
+
+func CmdDeactivateDid() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "deactivate-did [payload-json] [ver-method-id-1] [priv-key-1] [ver-method-id-N] [priv-key-N] ...",
```
Comment (Toktar) [inline x/cheqd/client/cli/tx_deactivate_did.go:13]: Why `payload-json`? It may be just "id"

File: x/cheqd/keeper/msg_server_deactivate_did.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+			return nil, err
+		}
+	}
+
+	// Verify signatures
+	updatedDid, err := updatedStateValue.UnpackDataAsDid()
+	if err != nil {
+		return nil, err
+	}
+	signers := GetSignerDIDsForDIDDeactivation(*existingDid, *updatedDid)
```
Comment (Toktar) [inline x/cheqd/keeper/msg_server_deactivate_did.go:65]: Why does this method use 2 params. Only existingDid is needed. `GetSignerDIDsForDIDDeactivation` should look like  `GetSignerDIDsForDIDCreation`

File: x/cheqd/keeper/msg_server_deactivate_did.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Apply changes: return original id and modify state
+	err = k.SetDid(&ctx, existingDid, &updatedMetadata)
+	if err != nil {
+		return nil, types.ErrInternal.Wrapf(err.Error())
+	}
```
Comment (Toktar) [inline x/cheqd/keeper/msg_server_deactivate_did.go:83]:
```suggestion
	err = VerifyAllSignersHaveAllValidSignatures(&k.Keeper, &ctx, inMemoryDids, msg.Payload.GetSignBytes(), signers, msg.Signatures)
	if err != nil {
		return nil, err
	}

	// Build state value
	value, err := types.NewStateValue(&existingDid, &updatedMetadata)
	if err != nil {
		return nil, err
	}

	err = k.SetDid(&ctx, &value)
	if err != nil {
		return nil, types.ErrInternal.Wrapf(err.Error())
	}
```
The better way

File: x/cheqd/keeper/msg_server_deactivate_did.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		Did:      existingDid,
+		Metadata: &updatedMetadata,
+	}, nil
+}
+
+func GetSignerDIDsForDIDDeactivation(existingDid types.Did, updatedDid types.Did) []string { 
+	res := existingDid.GetControllersOrSubject()
+	res = append(res, updatedDid.GetControllersOrSubject()...)
+	res = append(res, existingDid.GetVerificationMethodControllers()...)
+	res = append(res, updatedDid.GetVerificationMethodControllers()...)
```
Comment (Toktar) [inline x/cheqd/keeper/msg_server_deactivate_did.go:96]:
```suggestion
func GetSignerDIDsForDIDDeactivation(did types.Did) []string { 
        res := did.GetControllersOrSubject()
	res = append(res, did.GetVerificationMethodControllers()...)
```

File: x/cheqd/tests/deactivate_did_test.go [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+		BobKey1:      GenerateKeyPair(),
+		BobKey2:      GenerateKeyPair(),
+		BobKey3:      GenerateKeyPair(),
+		BobKey4:      GenerateKeyPair(),
+		CharlieKey1:  GenerateKeyPair(),
+		CharlieKey2:  GenerateKeyPair(),
+		CharlieKey3:  GenerateKeyPair(),
+		CharlieKey4:  GenerateKeyPair(),
+		ImposterKey1: GenerateKeyPair(),
+	}
```
Comment (Toktar) [inline x/cheqd/tests/deactivate_did_test.go:23]:
```suggestion
keys := GenerateTestKeys()
```

File: x/cheqd/tests/deactivate_did_test.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+		valid      bool
+		name       string
+		signerKeys []SignerKey
+		signers    []string
+		msg        *types.MsgDeactivateDidPayload
+		errMsg     string
+	}{
+		{
+			valid: true,
+			name:  "Valid: Key rotation works",
```
Comment (Toktar) [inline x/cheqd/tests/deactivate_did_test.go:35]: Key rotation is impossible for Deactivation tx. Changing the test name is needed

File: x/cheqd/tests/deactivate_did_test.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+		BobKey3:      GenerateKeyPair(),
+		BobKey4:      GenerateKeyPair(),
+		CharlieKey1:  GenerateKeyPair(),
+		CharlieKey2:  GenerateKeyPair(),
+		CharlieKey3:  GenerateKeyPair(),
+		CharlieKey4:  GenerateKeyPair(),
+		ImposterKey1: GenerateKeyPair(),
+	}
+
+	cases := []struct {
```
Comment (Toktar) [inline x/cheqd/tests/deactivate_did_test.go:25]: Could you add negative tests too?

File: x/cheqd/types/tx_msg_deactivate_did_payload.go [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+func (msg *MsgDeactivateDidPayload) ToDid() Did {
+	return Did{
+		Id: msg.Id,
+	}
+}
+
+// Validation
+
+func (msg MsgDeactivateDidPayload) Validate(allowedNamespaces []string) error {
+	return msg.ToDid().Validate(allowedNamespaces)
```
Comment (Toktar) [inline x/cheqd/types/tx_msg_deactivate_did_payload.go:18]:
```suggestion
	return validation.ValidateStruct(&msg,
		validation.Field(&msg.Id, validation.Required, IsDID(allowedNamespaces)),
	        )
```
+ add `validation "github.com/go-ozzo/ozzo-validation/v4"` to `import`
+ remove `ToDid` method

**Comment by Tweeddalex** [comment]
Hey @Toktar it would be good to understand what the status is here. Is the deactivate DID feature complete now? What are the next steps?

**Review by askolesov — APPROVED** [review]
Looks great! Thanks.

**Comment by askolesov** [comment]
Does anybody know why my commits don't trigger pipelines?

**Comment by askolesov** [comment]
Ok, got it, because it's a fork. We need a way to trigger them befor merge.

**Review by askolesov — CHANGES_REQUESTED** [review]
We need to trigger pipelines

**Review by Toktar — APPROVED** [review]

**Comment by ankurdotb** [comment]
Task linked: [DEV-1774 Merge DID Deactivation PRs](https://app.clickup.com/t/6600954/DEV-1774)

**Review by askolesov — APPROVED** [review]

**Pull request merged**